SIPS/sip-83.md

```diff
@@ -47,55 +71,111 @@ allows subsequent minting and burning transactions to rely on the calculated deb
+Utilising a debt pool snapshot could expose frontminting opportunities when there are large price shifts observed on the
+debt pool's underlying synths such as the ETH / BTC prices moving.
+It is worth noting that frontminting mitigation strategies were implemented in [SIP40](./sip-40.md) in the form of a
+24-hours mint and burn lock. This value can be increased via an SCCP to `48+ hours` to further reduce the frontminting
+opportunities until continuous rewards are implemented.
 
-The bot could also monitor the mempool for upcoming prices and pre-calculate the new `totalIssuedSynths` values to detect if it needs to update the debt pool snapshot after.
+However, to ensure that such front-minting opportunities are minimised, and that the debt deviation keeper needs to
+execute as infrequently as possible, the debt will also be adjusted whenever synth supplies change.
 
-#### Liveness
+This is simple in the case of minting and burning; when these operations occur, the debt pool snapshot will be updated
+with the amount of sUSD that is issued or burned as the issuance / burning of sUSD debt will change the
+debt pool size but not require latest prices to be retrieved.
 
-To ensure that the debt pool snapshot is not stale, there will be a staleness check on the last `timestamp` of when the debt pool snapshot was updated. Minting, burning and claiming rewards will revert if the debt pool snapshot is stale until it has been updated.
+When exchanges between non-sUSD synths occur, the prices are already retrieved to
+perform the exchange, but we will additionally retrieve the post-exchange total supplies of the involved non-sUSD
+synths to compute their total sUSD-denominated value. The delta between these value and the previously-computed ones
+will be applied to the total debt snapshot, and the new values cached to be used next time an exchange of those
+synths occurs.
+
+In this way the current debt snapshot will be responsive to the latest price updates of those synths which are most
+frequently used, and the public keeper function will ideally only need to be invoked at times of extreme synth price
+volatility or heavy network congestion. This system will assist in protecting against front minting attacks,
+but if it proves to keep the debt snapshot accurate enough, then its constant-time execution cost implies
+that an unbounded number of synths could be added to the system.
+
+#### Amortised Minter Debt Recomputation
+
+The debt snapshot keeper is the least technically complex solution to the debt calculation problem posed in this
+SIP, and is a useful safety mechanism for protocol participants to protect the system if required, but ideally
+no keeper functions would be required to maintain the integrity of the Synthetix debt pool.
+
+Therefore at a later stage it will be expedient to allow a full debt snapshot to occur at periodic heartbeats at the
```

Comment:
This could be set out more clearly / titled as future Phase 2 for periodic (~15mins-30mins) heartbeat snapshots to be taken and rebated with Ether from the gas tank?  Did we decide if the heartbeat is going to be executed during a minting, burning or claiming rewards transaction that falls on the heartbeat or a separate `rewardable` public function that can be called?   

The levy could be on every mint, burn, and claim reward operation (as 800k+ gas savings for these tx's)  ?

Reply:
I've clarified these points. As it's a phase 2 detail, I've left it open in the SIP whether it's a public function or triggered by mints/burns/claims, but flagged specifically that it could be either.